AudioDecoderCocoa should trim start of decoded samples in accordance with preSkip information

jyavenard opened the pull request:
#### 2b04af10a098c18c19cc2bc578ec6539c0087f26
<pre>
AudioDecoderCocoa should trim start of decoded samples in accordance with preSkip information
<a href="https://bugs.webkit.org/show_bug.cgi?id=284354">https://bugs.webkit.org/show_bug.cgi?id=284354</a>
<a href="https://rdar.apple.com/141202697">rdar://141202697</a>

Reviewed by Youenn Fablet.

Opus provides out of band information related to the encoder delay (a.k.a preSkip)
available in the AudioConfig&apos;s description.
Other codecs do no, so we could have assumed as AudioToolbox does, default preSkip
values (2112 for AAC, 536 for Lame MP4, 0 for Flac and an entire packet for vorbis).
However doing so, while technically more accurate, would cause WPT&apos;s test failures
as some existing mp3 and aac frames make strong assumptions on how many decoded frames
we return.

We use the preSkip, when available and set the AudioSampleBufferConverter to trim those frames.

- We also store the AudioSampleBufferConverter&apos;s Config object rather than individual values
as this configuration object will need to be extensively amended to support the future AudioEncoder object.
- Fly by: Return error if we failed to create and configure the AudioConverter.

Following this changes, we provide more accurate Opus decoding. There&apos;s no change in the result
of existing AudioDecoder&apos;s tests, but incoming AudioEncoder ones do rely on the files to be correctly
trimmed.

* Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
(WebCore::InternalAudioDecoderCocoa::initialize): Fly-by: Do not incorrectly reject Opus stereo content.
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.h:
(WebCore::AudioSampleBufferConverter::preSkip const):
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
(WebCore::m_options):
(WebCore::AudioSampleBufferConverter::initAudioConverterForSourceFormatDescription): Return error code when failing.
Change fallback rate from 44.1kHz to 48kHz. 48kHz is a more logical choice, it&apos;s supported by all hardware and is the default
for a codec (44.1kHz would have required to upsample to 48kHz first).
(WebCore::AudioSampleBufferConverter::attachPrimingTrimsIfNeeded):
(WebCore::AudioSampleBufferConverter::provideSourceDataNumOutputPackets):
(WebCore::AudioSampleBufferConverter::processSampleBuffers):
(WebCore::AudioSampleBufferConverter::bitRate const):
(WebCore::m_generateTimestamp): Deleted.
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::PacketDurationParser::framesInPacket):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h: Have parseOpusPrivateData take a span instead of a SharedBuffer
and return optional&lt;OpusCookieContent&gt; instead of using an out parameter.
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::parseOpusTOCData):
(WebCore::parseOpusPrivateData):
(WebCore::createOpusPrivateData):

Canonical link: <a href="https://commits.webkit.org/287708@main">https://commits.webkit.org/287708@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2637d4184871bd01511eb8c15fd4ca47540923f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80570 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59577 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34503 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85091 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31552 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82681 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68638 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7882 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62941 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20748 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53055 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73343 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43246 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50350 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27501 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30011 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71496 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28024 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86525 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7796 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71234 "Found 12 new test failures: imported/w3c/web-platform-tests/css/css-view-transitions/iframe-old-has-scrollbar.html imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-below-viewport-offscreen-old.html imported/w3c/web-platform-tests/css/css-view-transitions/names-are-tree-scoped.html imported/w3c/web-platform-tests/css/css-view-transitions/navigation/root-and-nested-element-transition.html imported/w3c/web-platform-tests/css/css-view-transitions/old-content-intrinsic-aspect-ratio.html imported/w3c/web-platform-tests/css/css-view-transitions/old-content-root-scrollbar-with-fixed-background.html imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-empty-iframe.html webanimations/accelerated-animation-opacity-animation-begin-time-after-scale-animation-ends.html webanimations/accelerated-transform-animation-to-scale-zero-with-implicit-from-keyframe.html webanimations/accelerated-transition-by-removing-property.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7971 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69180 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70475 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17565 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14481 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13424 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7758 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7597 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/11116 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9402 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
```diff
@@ -206,26 +205,26 @@
-        if (m_outputBitRate) {
-            if (auto error = PAL::AudioConverterSetProperty(m_converter, kAudioConverterEncodeBitRate, sizeof(*m_outputBitRate), &m_outputBitRate.value()))
+        if (m_options.outputBitRate) {
+            if (auto error = PAL::AudioConverterSetProperty(m_converter, kAudioConverterEncodeBitRate, sizeof(*m_options.outputBitRate), &m_options.outputBitRate.value()))
```
Nit: it is weird to see both `sizeof(*m_options.outputBitRate)` and `m_options.outputBitRate.value()` used here

youennf replied:
Right, how about `m_options->`?

jyavenard replied:
> Right, how about `m_options->`?

m_options isn't an optional, it doesn't have an `operator->()`

jyavenard replied:
> Nit: it is weird to see both `sizeof(*m_options.outputBitRate)` and `m_options.outputBitRate.value()` used here

getting the address is a different thing and looks less puzzling than &*m_options.outputBitRate IMHO

jyavenard replied:
I changed it to `&*` then

eric-carlson commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
```diff
@@ -235,16 +234,32 @@
+        UInt32 size = sizeof(primeInfo);
+        if (auto error = PAL::AudioConverterGetProperty(m_converter, kAudioConverterPrimeInfo, &size, &primeInfo))
+            RELEASE_LOG_ERROR(MediaStream, "AudioSampleBufferConverter getting kAudioConverterPrimeInfo failed with %d", static_cast<int>(error));
+        m_remainingPrimeDuration = PAL::CMTimeMake(primeInfo.leadingFrames, m_destinationFormat.mSampleRate);
```
Do we really want to set `m_remainingPrimeDuration` when the call to `AudioConverterGetProperty` failed?

jyavenard replied:
> Do we really want to set `m_remainingPrimeDuration` when the call to `AudioConverterGetProperty` failed?

Yes, it will be set to 0.
That's how we can then simply assess that m_remainingPrimeDuration is never invalid

youennf commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.h:
```diff
@@ -46,6 +46,7 @@ class AudioSampleBufferConverter : public ThreadSafeRefCountedAndCanMakeThreadSa
         std::optional<AudioStreamBasicDescription> description { };
         std::optional<unsigned> outputBitRate { };
         bool generateTimestamp { true };
+        std::optional<unsigned> preSkip { 0 };
```
No need for optional and 0. I would go with an unsigned { 0 } here.

jyavenard replied:
> No need for optional and 0. I would go with an unsigned { 0 } here.

no. They aren't the same thing.

There's absolutely a need for it.
The AudioConverter will by default trim some frames: 2112 with AAC, 576 with MP3, `SampleRate / 400` if Opus.
An empty optional indicate that we will let AudioConverter do the default and that's how we use it with the MediaRecorder, it provides better A/V sync that way.

A value of 0 indicates we don't want any trimming, which is how the AudioDecoder is using it and the WPT currently (incorrectly) tests that no frames are trimmed.

jyavenard replied:
I've cleared the default value, you were correct than as it was, it was unnecessary. But that's not how I am actually using it.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
```diff
@@ -179,18 +178,18 @@
-    if (error == 'fmt?' && outputFormatID != kAudioFormatOpus) {
-        m_destinationFormat.mSampleRate = 44100;
+    if (error == kAudioConverterErr_FormatNotSupported) {
+        m_destinationFormat.mSampleRate = 48000;
```
Maybe describe this change in commit log.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.h:
```diff
@@ -106,8 +107,8 @@ class AudioSampleBufferConverter : public ThreadSafeRefCountedAndCanMakeThreadSa
-    const std::optional<unsigned> m_outputBitRate;
-    const bool m_generateTimestamp { true };
+    const Options m_options;
+    std::atomic<unsigned> m_defaultBitRate { 0 };
```
Do we need an atomic or can we just use WTF_GUARDED_BY_CAPABILITY(queue().get())?

jyavenard replied:
It is set on the queue() and is read on the main thread by the MediaRecorder. So it needs to be atomic

jyavenard replied:
Note that this value is used where we havent' set a bitrate and we want to know what the AudioConverter chose for us.

youennf commented on Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
```diff
@@ -254,7 +254,7 @@ bool parseOpusTOCData(const SharedBuffer& frameData, OpusCookieContents& cookie)
 }
 
-bool parseOpusPrivateData(std::span<const uint8_t> codecPrivateData, SharedBuffer& frameData, OpusCookieContents& cookie)
+bool parseOpusPrivateData(std::span<const uint8_t> codecPrivateData, std::span<const uint8_t> frameData, OpusCookieContents& cookie)
```
Pre-exisiting issue, but it is unclear whether only part of cookie values will be updated here.
We can probably just send an std::optional< OpusCookieContents> instead.

jyavenard replied:
OpusCookieContents is an out parameter, I don't see why you would want to make this an optional.

youennf replied:
I was thinking of:
```suggestion
std::optional<OpusCookieContents > parseOpusPrivateData(std::span<const uint8_t> codecPrivateData, std::span<const uint8_t> frameData)
```

jyavenard replied:
I see.
good idea

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -248,9 +249,10 @@ String InternalAudioDecoderCocoa::initialize(const String& codecName, const Audi
     if (!format) {
-        if (codec == kAudioFormatOpus && config.numberOfChannels >= 2)
+        if (codec == kAudioFormatOpus && config.numberOfChannels > 2)
             return "Opus with more than 2 channels isn't supported"_s;
```
We should add a test for this one. I guess there might already be a WPT but if not, let's add one.

jyavenard replied:
yes, there's a test in the audio-encoder where it encodes and then decode; the content is stereo.

youennf commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -1276,7 +1276,7 @@ WebMParser::ConsumeFrameDataResult WebMParser::AudioTrackData::consumeFrameData(
     }
 
-    MediaTime packetDuration = MediaTime(m_packetDurationParser->framesInPacket(*contiguousBuffer), downcast<AudioInfo>(formatDescription())->rate);
+    MediaTime packetDuration = MediaTime(m_packetDurationParser->framesInPacket(contiguousBuffer->span()), downcast<AudioInfo>(formatDescription())->rate);
```
```suggestion
    MediaTime packetDuration { m_packetDurationParser->framesInPacket(contiguousBuffer->span()), downcast<AudioInfo>(formatDescription())->rate };
```

jyavenard replied:
That requires static_cast making the expression even longer
`Non-constant-expression cannot be narrowed from type 'size_t' (aka 'unsigned long') to 'int64_t' (aka 'long long') in initializer list`

jyavenard committed:
AudioDecoderCocoa should trim start of decoded samples in accordance with preSkip information
https://bugs.webkit.org/show_bug.cgi?id=284354
rdar://141202697

Reviewed by Youenn Fablet.

Opus provides out of band information related to the encoder delay (a.k.a preSkip)
available in the AudioConfig's description.
Other codecs do no, so we could have assumed as AudioToolbox does, default preSkip
values (2112 for AAC, 536 for Lame MP4, 0 for Flac and an entire packet for vorbis).
However doing so, while technically more accurate, would cause WPT's test failures
as some existing mp3 and aac frames make strong assumptions on how many decoded frames
we return.

We use the preSkip, when available and set the AudioSampleBufferConverter to trim those frames.

- We also store the AudioSampleBufferConverter's Config object rather than individual values
as this configuration object will need to be extensively amended to support the future AudioEncoder object.
- Fly by: Return error if we failed to create and configure the AudioConverter.

Following this changes, we provide more accurate Opus decoding. There's no change in the result
of existing AudioDecoder's tests, but incoming AudioEncoder ones do rely on the files to be correctly
trimmed.

* Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
(WebCore::InternalAudioDecoderCocoa::initialize): Fly-by: Do not incorrectly reject Opus stereo content.
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.h:
(WebCore::AudioSampleBufferConverter::preSkip const):
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
(WebCore::m_options):
(WebCore::AudioSampleBufferConverter::initAudioConverterForSourceFormatDescription): Return error code when failing.
Change fallback rate from 44.1kHz to 48kHz. 48kHz is a more logical choice, it's supported by all hardware and is the default
for a codec (44.1kHz would have required to upsample to 48kHz first).
(WebCore::AudioSampleBufferConverter::attachPrimingTrimsIfNeeded):
(WebCore::AudioSampleBufferConverter::provideSourceDataNumOutputPackets):
(WebCore::AudioSampleBufferConverter::processSampleBuffers):
(WebCore::AudioSampleBufferConverter::bitRate const):
(WebCore::m_generateTimestamp): Deleted.
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::PacketDurationParser::framesInPacket):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h: Have parseOpusPrivateData take a span instead of a SharedBuffer
and return optional<OpusCookieContent> instead of using an out parameter.
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::parseOpusTOCData):
(WebCore::parseOpusPrivateData):
(WebCore::createOpusPrivateData):

Canonical link: https://commits.webkit.org/287708@main
